nfd-gc: Remove stale NRT objects

ozhuraki opened the pull request:
Remove stale NRT objects whose creator pod does not exist anymore.

Closes #1586

ArangoGutierrez commented:
/ok-to-test

PiotrProkop commented on pkg/nfd-gc/nfd-gc.go:
```diff
@@ -146,6 +146,16 @@ func (n *nfdGarbageCollector) garbageCollect() {
+	}
+	podNames := sets.NewString()
+	for _, pod := range pods {
+		nodeNames.Insert(pod.Name)
```
you are inserting into `nodeNames` ?

PiotrProkop commented on pkg/nfd-gc/nfd-gc.go:
```diff
@@ -146,6 +146,16 @@ func (n *nfdGarbageCollector) garbageCollect() {
 		nodeNames.Insert(node.Name)
 	}
 
+	pods, err := n.factory.Core().V1().Pods().Lister().List(labels.Everything())
```
are you listing all Pods in the cluster?

PiotrProkop commented on pkg/nfd-gc/nfd-gc.go:
```diff
@@ -175,6 +185,9 @@ func (n *nfdGarbageCollector) garbageCollect() {
 			if !nodeNames.Has(nrt.Name) {
 				n.deleteNRT(nrt.Name)
 			}
+			if !podNames.Has(nrt.Labels["pod-name"]) {
```
why `pod-name` value? use const?

ArangoGutierrez commented on pkg/nfd-topology-updater/nfd-topology-updater.go:
```diff
@@ -294,6 +294,7 @@ func (w *nfdTopologyUpdater) updateNodeResourceTopology(zoneInfo v1alpha2.ZoneLi
 		nrtNew := v1alpha2.NodeResourceTopology{
 			ObjectMeta: metav1.ObjectMeta{
 				Name:            w.nodeName,
+				Labels:          map[string]string{"owner-pod": os.Getenv("POD_NAME")},
```
Let's use https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L295 that is the to-go way to define ownership instead of a custom label

ArangoGutierrez replied:
My bad as @PiotrProkop commented to my offline, you can not set OwnerReference to a Cluster scoped object

PiotrProkop commented on pkg/nfd-topology-updater/nfd-topology-updater.go:
```diff
@@ -294,6 +294,7 @@ func (w *nfdTopologyUpdater) updateNodeResourceTopology(zoneInfo v1alpha2.ZoneLi
 		nrtNew := v1alpha2.NodeResourceTopology{
 			ObjectMeta: metav1.ObjectMeta{
 				Name:            w.nodeName,
+				Labels:          map[string]string{"owner-pod": os.Getenv("POD_NAME")},
```
why `owner-pod` ? use const? Pod names are only unique in the namespace not across whole K8s cluster.

ozhuraki commented:
@PiotrProkop @ArangoGutierrez

Thanks! Updated, please take a look

ozhuraki commented:
@PiotrProkop @ArangoGutierrez

Updated unit tests, please take a look

PiotrProkop commented:
@ozhuraki @ArangoGutierrez one question, if we upgrade NFD-gc on existing cluster with this feature, it would remove all NRTs in the cluster cause none of them, has the `owner-pod` labels? Is it acceptable? Even if NRTs would be recreated after some time automatically it can lead to  some downtime for components leveraging NRTs like topo-aware-scheduler.
/cc @ffromani

ozhuraki commented:
@PiotrProkop
Thanks for the input. Yes, this would happen.

I updated it so that NRT is garbage collected only when it has label, please take a look.

marquiz commented on pkg/nfd-gc/nfd-gc.go:
```diff
@@ -37,6 +38,10 @@ import (
 )
 
+const (
+	OwnerPod = "owner-pod"
```
I think the label should have a prefix

```suggestion
	NRTOwnerPodLabel = "nfd.node.kubernetes.io/owner-pod"
```

marquiz commented on pkg/nfd-topology-updater/nfd-topology-updater.go:
```diff
@@ -294,6 +295,8 @@ func (w *nfdTopologyUpdater) updateNodeResourceTopology(zoneInfo v1alpha2.ZoneLi
 		nrtNew := v1alpha2.NodeResourceTopology{
 			ObjectMeta: metav1.ObjectMeta{
 				Name:            w.nodeName,
+				Namespace:       w.kubernetesNamespace,
```
NRT is is cluster-scoped so it doesn't have a namespace

```suggestion
```

marquiz commented on pkg/nfd-topology-updater/nfd-topology-updater.go:
```diff
@@ -294,6 +295,8 @@ func (w *nfdTopologyUpdater) updateNodeResourceTopology(zoneInfo v1alpha2.ZoneLi
 			ObjectMeta: metav1.ObjectMeta{
 				Name:            w.nodeName,
+				Namespace:       w.kubernetesNamespace,
+				Labels:          map[string]string{nfdgc.OwnerPod: os.Getenv("POD_NAME")},
```
I think we should include the namespace in the label (and then check it in nfd-gc)

```suggestion
				Labels:          map[string]string{nfdgc.OwnerPod: fmt.Sprintf("%s/%s", w.kubernetesNamespace, os.Getenv("POD_NAME"))},
```

marquiz reviewed (commented):
Thanks @ozhuraki for the PR. A few comments from me below.

@PiotrProkop PTAL

ozhuraki commented:
@marquiz 
Thanks! Updated, please take a look

marquiz commented on pkg/nfd-gc/nfd-gc.go:
```diff
@@ -172,7 +180,26 @@ func (n *nfdGarbageCollector) garbageCollect() {
+			}
+
+			if label, ok := nrt.Labels[NRTOwnerPodLabel]; ok {
+				if s := strings.Split(string(label), "/"); len(s) == 2 {
```
nit: no need for type conversion

```suggestion
				if s := strings.Split(label, "/"); len(s) == 2 {
```

marquiz commented on pkg/nfd-topology-updater/nfd-topology-updater.go:
```diff
@@ -294,6 +295,7 @@ func (w *nfdTopologyUpdater) updateNodeResourceTopology(zoneInfo v1alpha2.ZoneLi
 		nrtNew := v1alpha2.NodeResourceTopology{
 			ObjectMeta: metav1.ObjectMeta{
 				Name:            w.nodeName,
+				Labels:          map[string]string{nfdgc.NRTOwnerPodLabel: fmt.Sprintf("%s/%s", w.kubernetesNamespace, os.Getenv("POD_NAME"))},
```
We should also set the label when updating the NRT. Otherwise we get into trouble with pod restarts/updates (when the name changes) because gc doesn't find the pod and GCs the NRT object.

marquiz commented on pkg/nfd-gc/nfd-gc.go:
```diff
@@ -37,6 +39,10 @@ import (
 )
 
+const (
+	NRTOwnerPodLabel = "nfd.node.kubernetes.io/owner-pod"
```
nit: I think the "owner" of the label name is the nfd-topology-updater so this would be better to move in `pkg/nfd-topology-updater/nfd-topology-updater.go`. Makes sense?

marquiz commented on pkg/nfd-gc/nfd-gc.go:
```diff
@@ -172,7 +180,26 @@ func (n *nfdGarbageCollector) garbageCollect() {
+				if s := strings.Split(string(label), "/"); len(s) == 2 {
+					ns := s[0]
+					pod := s[1]
+					_, err := n.k8sClient.CoreV1().Pods(ns).Get(context.TODO(), pod, metav1.GetOptions{})
```
Hmm, I think there's a window for nasty race condition in case of (topology-updater) pod restarts:

a) topology-updater pod is deleted
b) nfd-gc kicks in and doesn't see the pod (even if it's temporarily gone)
c) nfd-gc GCs the NRT
d) -> problems 

One possible solution would be to have keep a list (e.g. `map[string]string`, key is the NRT name, value is the pod ns/name) of stale NRTs. On the first GC round, if the topology-updater pod is not found mark the NRT on this list. On the next GC round, if the pod is still not found (and the pod ns/name hasn't changed) delete the NRT. WDYT?

ping @PiotrProkop @ffromani

PiotrProkop replied:
I think we'll end up with this race condition every time when nfd-topology-updater will be upgraded in the cluster, as the name of the pod(created by daemonset) will change on the node.
Another idea would be to somehow check if `nfd-topology-updater` is being handled by some controller(via OwnerReference) and just check if this controller still exists? 
I am still debating if I like this idea or not as it may make the code much more complicated and I don't know if it's worth it 😄 even though 99% of the deployment should create `nfd-topology-updater` via daemon-sets we can't guarantee that, so we would need to also write the code for this 1% of deployments.

marquiz replied:
With what I suggested the race should be solved. Don't GC immediately but wait for two GC periods. If the version (or pod template spec in the DaemonSet whatever) is updated the new pod will update the owner pod label -> the NRT would not be GC'd, even if the new pod would be (temporarily) gone, too. Only when the owner pod (marked by the special label, we could use annotation, too) is gone for two consecutive GC periods the NRT would be GC'd.

This method doesn't depend on how the topology-updater is deployed (daemonset, deployment, custom controller) as the "owner ref" label is set by the topology-updater pod.

marquiz reviewed (commented):
Thanks for the update @ozhuraki. I identified a few possible race issues but those shouldn't be too hard to solve.

ozhuraki commented:
@marquiz @PiotrProkop

Thanks for the help here and apologies, missed your comments.

Updated per your suggestion to solve the race and rebased, please take a look

marquiz commented on pkg/nfd-gc/nfd-gc.go:
```diff
@@ -190,7 +198,27 @@ func (n *nfdGarbageCollector) garbageCollect() {
+				if errors.IsNotFound(err) {
+					if _, ok := gcNRTs[meta.Name]; !ok {
+						gcNRTs[meta.Name] = label
+					} else {
```
Wouldn't we need to check the value as well? `val == label`. If the pod name (or namespace) has changed, DON'T delete the NRT.

I just realized that the `label` is somewhat misleading name for the variable. How about e.g. `objRef` or `objKey` or smth?

marquiz commented on pkg/nfd-gc/nfd-gc.go:
```diff
@@ -57,10 +62,11 @@ type NfdGarbageCollector interface {
+	stopChan  chan struct{}
+	client    metadataclient.Interface
+	factory   metadatainformer.SharedInformerFactory
+	k8sClient k8sclient.Interface
```
We shouldn't need a separate client but just reuse the metadataclient above (we're not interested in the content of the podspec)

```suggestion
```

ozhuraki commented:
@marquiz

Thanks for the useful input. Updated, please take a look
